Frends.FTP.UploadFiles/Frends.FTP.UploadFiles/Definitions/SingleFileTransfer.cs

```diff
@@ -0,0 +1,379 @@
+                if (_batchContext.Options.RenameSourceFileBeforeTransfer)
+                    RenameSourceFile();
+                else 
+                    _sourceFileDuringTransfer = SourceFile.FullPath;
+
+                if (DestinationFileExists(DestinationFileNameWithMacrosExpanded))
+                {
+                    switch (_batchContext.Destination.Action)
+                    {
+                        case DestinationAction.Append:
+                            AppendDestinationFile();
+                            break;
+                        case DestinationAction.Overwrite:
+                            PutDestinationFile(removeExisting: true);
+                            break;
+                        case DestinationAction.Error:
+                            throw new DestinationFileExistsException(DestinationFileNameWithMacrosExpanded);
+                    }
+                }
+                else
+                {
+                    PutDestinationFile();
+                }
+
+                if (_batchContext.Options.PreserveLastModified)
+                    RestoreModified();
+
+                ExecuteSourceOperation();
+                _logger.LogTransferSuccess(this, _batchContext);
+                CleanUpFiles();
+            }
+            catch (Exception ex)
+            {
+                var sourceFileRestoreMessage = RestoreSourceFileAfterErrorIfItWasRenamed();
+                HandleTransferError(ex, sourceFileRestoreMessage);
+            }
+            CleanUpFiles();
+            return _result;
+        }
+
+        /// <summary>
+        /// Checks if destination file exists.
+        /// </summary>
+        /// <returns></returns>
+        private bool DestinationFileExists(string filename)
+        {
+            Trace(
+                TransferState.CheckIfDestinationFileExists,
+                "Checking if destination file {0} exists",
+                SourceFile.Name);
+            return _client.FileExists(filename);
+        }
+
+        /// <summary>
+        /// Rename file with unique file name. 
+        /// </summary>
+        private void RenameSourceFile()
+        {
+            var uniqueFileName = Util.CreateUniqueFileName();
+            var workdir = !string.IsNullOrEmpty(_batchContext.Info.WorkDir)
+                ? _batchContext.Info.WorkDir
+                : Path.GetDirectoryName(SourceFile.FullPath);
+            var renamedFile = Path.Combine(workdir, uniqueFileName);
+
+            Trace(TransferState.RenameSourceFileBeforeTransfer, "Renaming source file {0} to temporary file name {1} before transfer", SourceFile.Name, uniqueFileName);
+            File.Move(SourceFile.FullPath, renamedFile);
+
+            _sourceFileDuringTransfer = renamedFile;
+        }
+
+        /// <summary>
+        /// Appends source file content to existing destination file.
+        /// </summary>
+        private void AppendDestinationFile()
```

Comment:
I didn't see was there a rename done for the destination file before appending. I think in this it is crucial because the destination file will be used like wise in PutDestinationFile.

Reply:
Yes, the rename here is not handled at all. I added proper handling and added two units test for those things.